controllers/btpoperator_controller.go

```diff
@@ -424,75 +220,178 @@ func (r *BtpOperatorReconciler) HandleProcessingState(ctx context.Context, cr *v
-			return r.UpdateBtpOperatorStatus(ctx, cr, types.StateError, ProvisioningFailed, "Timed out while waiting for resources readiness")
+		if len(value) == 0 {
+			missingValues = append(missingValues, key)
 		}
 	}
+	if len(missingKeys) > 0 {
+		missingKeysMsg := fmt.Sprintf("key(s) %s not found", strings.Join(missingKeys, ", "))
+		errs = append(errs, missingKeysMsg)
+	}
+	if len(missingValues) > 0 {
+		missingValuesMsg := fmt.Sprintf("missing value(s) for %s key(s)", strings.Join(missingValues, ", "))
+		errs = append(errs, missingValuesMsg)
+	}
+	if len(errs) > 0 {
+		return fmt.Errorf("%s", strings.Join(errs, ", "))
+	}
 
-	logger.Info("provisioning succeeded")
-	return r.UpdateBtpOperatorStatus(ctx, cr, types.StateReady, ReconcileSucceeded, "Module provisioning succeeded")
+	return nil
 }
 
-func (r *BtpOperatorReconciler) createUnstructuredObjectsFromManifestsDir(ctx context.Context, manifestsDir string) ([]*unstructured.Unstructured, error) {
+func (r *BtpOperatorReconciler) deleteOutdatedResources(ctx context.Context) error {
 	logger := log.FromContext(ctx)
 
+	logger.Info("getting outdated module resources to delete")
+	usToDelete, err := r.createUnstructuredObjectsFromManifestsDir(r.getResourcesToDeletePath())
+	if err != nil {
+		logger.Error(err, "while getting deletable objects from manifests")
+		return fmt.Errorf("Failed to create deletable objects from manifests: %w", err)
+	}
+	logger.Info(fmt.Sprintf("got %d outdated module resources to delete", len(usToDelete)))
+
+	err = r.deleteResources(ctx, usToDelete)
+	if err != nil {
+		logger.Error(err, "while deleting outdated resources")
+		return fmt.Errorf("Failed to delete outdated resources: %w", err)
+	}
+
+	return nil
+}
+
+func (r *BtpOperatorReconciler) createUnstructuredObjectsFromManifestsDir(manifestsDir string) ([]*unstructured.Unstructured, error) {
 	objs, err := r.manifestHandler.CollectObjectsFromDir(manifestsDir)
 	if err != nil {
-		logger.Error(err, "while collecting objects from manifests")
 		return nil, err
 	}
 	us, err := r.manifestHandler.ObjectsToUnstructured(objs)
 	if err != nil {
-		logger.Error(err, "while converting objects to Unstructured")
 		return nil, err
 	}
 
 	return us, nil
 }
 
-func (r *BtpOperatorReconciler) prepareModuleResources(ctx context.Context, us []*unstructured.Unstructured) *ErrorWithReason {
+func (r *BtpOperatorReconciler) getResourcesToDeletePath() string {
+	return fmt.Sprintf("%s%cdelete", ResourcesPath, os.PathSeparator)
+}
+
+func (r *BtpOperatorReconciler) deleteResources(ctx context.Context, us []*unstructured.Unstructured) error {
+	logger := log.FromContext(ctx)
+
+	var errs []string
+	for _, u := range us {
+		if err := r.Delete(ctx, u); err != nil && !k8serrors.IsNotFound(err) {
+			errs = append(errs, fmt.Sprintf("failed to delete %s %s: %s", u.GetName(), u.GetKind(), err))
+		} else {
+			logger.Info("deleted resource", "name", u.GetName(), "kind", u.GetKind())
```

Comment:
Nit: in case there is no resource of given name/kind we log that it was deleted. But
- it could be deleted earlier for unknown reasons (e.g. manually by user)
- it could have never existed

Reply:
that's true, I missed that. Changed the if statement to continue the loop when the error is NotFound type.